### DISCUSSION TITLE
Question about engine-pair concept in Navy

### DISCUSSION THREAD

**Discussion by arungeorge83** [facebook/CacheLib#221] [Q&A]
I see that a concept of Engine pairs (multiple pairs of BigHash & BlockCache) is added into the initialization sequence of Cachelib.
https://github.com/facebook/CacheLib/blob/main/cachelib/allocator/nvmcache/NavySetup.cpp#L241 
And I don't see any documentation or much comments on the code about it.

Could you explain the main use cases of this concept?
Thanks

**Comment by haowu14** [comment]
Hi, the idea is similar to the "pools" in memory. If you have multiple use cases in the same cachelib instance, you can use pools to separate them in memory and engine pairs to separate them in NVM.

**Comment by arungeorge83** [comment]
Thanks.
A few questions more please.

1 - How does a 'key' or 'NVM IO' is directed to a particular navy engine pair? what is the mapping logic for the 'key -> engine pair' ?
2 - is there  any mapping between memory pool and 'navy engine pair'
3 - Are you guys planning to publish any 'cachebench/test_configs/' with engine pair examples?

**Reply by jaesoo-fb** [reply]
Hi @arungeorge83 

Let me answer the follow up questions.

1. 
The `EnginePair` is selected based on the policy provided in `NavyConfig` via `NavyConfig::setEnginesSelector`

2. The `EngineSelector` is a `std::function` taking `HashedKey` as input, meaning the `EnginePair` could be determined only based on the `HashedKey`.
```
using EnginesSelector = std::function<size_t(HashedKey)>;
```

3. No plan yet. Even the `EnginePair` is not supported in the cachebench's `CacheConfig` yet, either.

**Reply by arungeorge83** [reply]
Hey @jaesoo-fb @haowu14,
Thanks for the replies.

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
